fix: backward compatibility with numberofchunks

zaratsian opened the pull request:
What type of PR is this?
/kind fix

What this PR does / Why we need it:
This PR maps a legacy field called "NumberOfChunks" to the newly named "ChunkCount" field. This change was implemented against the record and blobref so that all downstream functions are able to reference "ChunkCount" without there being any naming conflict. 

Which issue(s) this PR fixes:  Backward compatibility issue with Record created with NumberOfChunks #391
https://github.com/googleforgames/open-saves/issues/391

Special notes for your reviewer:

hongalex commented:
/gcbrun

hongalex reviewed (commented):
Code itself looks good, could you create/modify a test with an entity that has `NumberOfChunks` and make sure this loading is done properly?

yuryu commented on internal/pkg/metadb/blobref/blobref.go:
```diff
@@ -71,7 +71,16 @@ func (b *BlobRef) Save() ([]datastore.Property, error) {
-	return datastore.LoadStruct(b, ps)
+        // ps_modified has been added to enable backward compatibility by creating 
+        // a new datastore array that replaces legacy "NumberOfChunks" with "ChunkCount". 
+        var ps_modified []datastore.Property
```
You can directly change `ps` I think. If you need to use a new slice, please use mixedCaps.

https://go.dev/doc/effective_go#names

loranger2k replied:
I checked in the history, and "NumberOfChunks" has never been a field of BlobRef. When introduced, it was called ChunkCount. You can remove this change.

yuryu commented on internal/pkg/metadb/record/record.go:
```diff
@@ -83,7 +83,16 @@ func (r *Record) Save() ([]datastore.Property, error) {
-	externalBlob, ps, err := timestamps.LoadUUID(ps, externalBlobPropertyName)
+        // ps_modified has been added to enable backward compatibility by creating 
+        // a new datastore array that replaces legacy "NumberOfChunks" with "ChunkCount". 
+        var ps_modified []datastore.Property
```
ditto

loranger2k replied:
Although this fixed the backward compatibility issue within Open Saves, we would still need to change our gRPC client to start passing ChunkCount instead of NumberOfChunks. Since we use a rolling deployment strategy, we need a gRPC client that can handle both fields simultaneously.

loranger2k replied:
Thinking more about this, this code change avoids the internal error we were seeing but won't really provide data consistency since we always read "ChunkCount" from datastore.

loranger2k replied:
I take this back. I didn't realize the data is already loaded at this point. @zaratsian Are you going to modify the implementation and add a test for this?

zaratsian replied:
That sounds good - I'm working on a few updates based on the feedback, plus the tests and will update the PR shortly.

yuryu reviewed (commented):
Overall lgtm, would probably need a few test cases.

yuryu commented:
@zaratsian fyi, please take what you need from here https://github.com/googleforgames/open-saves/compare/main...yuryu:open-saves:number-count

zaratsian commented:
> @zaratsian fyi, please take what you need from here [main...yuryu:open-saves:number-count](https://github.com/googleforgames/open-saves/compare/main...yuryu:open-saves:number-count)

Your code suggestions looked great @yuryu, thanks! I've tested several ways as well and all references to NumberOfChunks seems to be handled well. @loranger2k I incorporated your feedback as well, and removed the blobref. I agree that should not be needed because it was introduced after the PR #391.

loranger2k reviewed (approved):
looks good.
